### PR TITLE
imp(08-wasm): supersede `Bytes` with `cosmwasm_std::Binary`

### LINKED ISSUE #1271: Replace the `Bytes` alias in CosmWasm context with `cosmwasm_std::Binary`
## Improvement Summary

Currently, the CosmWasm context uses a `Bytes` alias to serialize/deserialize Base64 strings to bytes through the use of Serde tags (for example, [see here](https://github.com/cosmos/ibc-rs/blob/7ff41b3804cda649e0d7f7c21389b8d65f8d9b4b/ibc-clients/cw-context/src/types/msgs.rs#L23-L25)). However, Bytes can be replaced with `cosmwasm_std::Binary`, which allows us to remove all the Serde and Schemars tags. Here is an example of the proposed change:

```diff
#[cw_serde]
pub struct InstantiateMsg {
-    #[schemars(with = "String")]
-    #[serde(with = "Base64", default)]
-    pub client_state: Bytes,
-    #[schemars(with = "String")]
-    #[serde(with = "Base64", default)]
-    pub consensus_state: Bytes,
-    #[schemars(with = "String")]
-    #[serde(with = "Base64", default)]
-    pub checksum: Bytes,
+    pub client_state: cosmwasm_std::Binary,
+    pub consensus_state: cosmwasm_std::Binary,
+    pub checksum: cosmwasm_std::Binary,
}
```

## Context

As part of the ongoing migration in the [08-wasm contract API](https://github.com/cosmos/ibc-go/issues/6496), we recommend moving away from using the `Bytes` alias in favor of `cosmwasm_std::Binary`.

The recent changes to the merkle path in `08-wasm` can be seen [here](https://github.com/cosmos/ibc-go/pull/6769/files#diff-f096e5301c62d95cc77f44fbce0b278bc633d536f0625a94810200058c646ca6R7-R13). Until more generic paths are supported in ibc-rs (see [issue #1255](https://github.com/cosmos/ibc-rs/issues/1255)), these bytes must be asserted as UTF-8 strings before being passed down to the Tendermint context.

By making this change, we simplify the code and reduce potential points of failure related to serialization/deserialization.

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: #1271

______

### PR author checklist:
- [X] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [X] Linked to GitHub issue.
- [X] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: ibc-clients/cw-context/src/types/response.rs [hunk 1]
```diff
@@ -1,12 +1,12 @@
 //! Contains the response types for the CosmWasm contract.
 use cosmwasm_schema::cw_serde;
-use ibc_client_wasm_types::Bytes;
```
Comment (rnbguy) [inline ibc-clients/cw-context/src/types/response.rs:3]: Ideally, we should remove the `ibc_client_wasm_types::Bytes` altogether. I tried to do that and noticed, our `Bytes` are being used at other places too.

PS. Probably, you are still working on the PR. In that case, ignore this comment.

File: tests-integration/tests/cosmwasm/helper.rs [hunk 2]
```diff
@@ -13,9 +13,10 @@ pub fn dummy_msg_info() -> MessageInfo {
     message_info(&creator, &coins(1000, "ibc"))
 }
 
-pub fn dummy_checksum() -> Vec<u8> {
+pub fn dummy_checksum() -> Binary {
```
Comment (rnbguy) [inline tests-integration/tests/cosmwasm/helper.rs]: can we use `Checksum` type alias here?

File: ibc-clients/cw-context/src/types/response.rs (hunk 1 above)
Comment (Farhad-Shabani) [reply ibc-clients/cw-context/src/types/response.rs:3]: as talked and commented [here](https://github.com/cosmos/ibc-rs/issues/1271#issuecomment-2214573623) to be able to use 08-wasm types in no_std envs, we can't quite take a consistent approach at the moment.

File: tests-integration/tests/cosmwasm/helper.rs (hunk 2 above)
Comment (Farhad-Shabani) [reply tests-integration/tests/cosmwasm/helper.rs]: I ended up using `cosmwasm_std::Checksum`
[8181b44](https://github.com/cosmos/ibc-rs/pull/1272/commits/8181b44a01d619d7a946f2390ae2b36fdec8c6d2)

**Review by rnbguy — APPROVED** [review]
LGTM :pray: Thanks Farhad !

File: .changelog/unreleased/breaking-changes/1271-supersede-Bytes-with-cosmwasm-Binary.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+- [ibc-client-cw] Supersede `Bytes` with `cosmwasm::Binary`.
+  ([\#1271](https://github.com/cosmos/ibc-rs/issues/1271))
```
Comment (rnbguy) [inline .changelog/unreleased/breaking-changes/1271-supersede-Bytes-with-cosmwasm-Binary.md]: One last comment. Why do you think this is a breaking change? The way I see, the users of our cw-context shouldn't care about these changes. For example, we didn't need to touch our Tendermint light client cw contract.

If you agree, I would suggest putting this under improvements.

Comment (Farhad-Shabani) [reply .changelog/unreleased/breaking-changes/1271-supersede-Bytes-with-cosmwasm-Binary.md]: Ah, that was left over from the first commit, as I had touched ibc-client-wasm-types as well. fixed [9515258](https://github.com/cosmos/ibc-rs/pull/1272/commits/9515258ae09a79cf68d58ea428024549b2583612)

**Pull request merged**